transport/tlscommon/diag.go

```diff
@@ -0,0 +1,178 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package tlscommon
+
+import (
+	"bytes"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"log"
+	"time"
+)
+
+// DiagCerts returns a diagnostics hook callback that will validate if the certifiactes (cert + key, and CAs) present in the config are valid.
+func (c *Config) DiagCerts() func() []byte {
+	if c == nil {
+		return func() []byte {
+			return []byte("error: nil tlscommon.Config\n")
+		}
+	}
+	return func() []byte {
+		var b bytes.Buffer
+		logger := log.New(&b, "tlscommon.Config: ", 0)
+		logger.Printf("Start diagnostics %s", time.Now().UTC())
+		logger.Printf("verification_mode=%s", c.VerificationMode)
+		logger.Printf("ca_trusted_fingerprint=%s", c.CATrustedFingerprint)
+		logger.Printf("ca_sha256=%v", c.CASha256)
+
+		diagCertificate(logger, &c.Certificate)
+		diagCAs(logger, c.CAs)
+
+		return b.Bytes()
+	}
+}
+
+// DiagCerts returns a diagnostics hook callback that will validate if the certifiactes (cert + key, and CAs) present in the config are valid.
+//
+// Implementation is mostly a copy of Config.DiagCerts
+func (c *ServerConfig) DiagCerts() func() []byte {
+	if c == nil {
+		return func() []byte {
+			return []byte("error: nil tlscommon.ServerConfig\n")
+		}
+	}
+	return func() []byte {
+		var b bytes.Buffer
+		logger := log.New(&b, "tlscommon.ServerConfig: ", 0)
+		logger.Printf("Start diagnostics %s", time.Now().UTC())
+		logger.Printf("verification_mode=%s", c.VerificationMode)
+		logger.Printf("client_auth=%s", c.ClientAuth)
+		logger.Printf("ca_sha256=%v", c.CASha256)
+
+		diagCertificate(logger, &c.Certificate)
+		diagCAs(logger, c.CAs)
+
+		return b.Bytes()
```

Comment:
From the fleet-server demo, output is:

```
tlscommon.ServerConfig: Start diagnostics 2024-05-29 00:18:41.569890423 +0000 UTC
tlscommon.ServerConfig: verification_mode=full
tlscommon.ServerConfig: client_auth=none
tlscommon.ServerConfig: ca_sha256=[]
tlscommon.ServerConfig: CertificateSettings: checking certificate keypair
tlscommon.ServerConfig: CertificateSettings: certificate keypair OK.
tlscommon.ServerConfig: CertificateSettings: cert 0 - Subject: CN=fleet-server-dev,O=elastic-fleet
	Issuer: CN=localhost,O=elastic-fleet
	NotBefore: 2024-05-29 00:18:15 +0000 UTC
	NotAfter: 2034-05-29 00:18:15 +0000 UTC
	Fingerprint: EvAr8erNJZJ0BoXor1jnCl5X2WCU79ISMgwMITOEJBs=
	SAN IP: []
	SAN DNS: [fleet-server-dev]
tlscommon.ServerConfig: CertificateAuthorities: certificate_authorities provided.
tlscommon.ServerConfig: CertificateAuthorities: - cert 0 Subject: CN=localhost,O=elastic-fleet
	IsCa: true
	BasicConstraintsValid: true
	NotBefore: 2024-05-29 00:18:15 +0000 UTC
	NotAfter: 2034-05-29 00:18:15 +0000 UTC
	Fingerprint: cSawoNEa2cSPV4Mwce4mRysrcBGtHBGcSeoBL+Cr48Q=
```

We want to make sure that certificate & ca information is provided in the diagnostics

Reply:
Slight update after I added `CertDiagString`
```
tlscommon.ServerConfig: Start diagnostics 2024-05-29 22:10:07.333735518 +0000 UTC
tlscommon.ServerConfig: verification_mode=full
tlscommon.ServerConfig: client_auth=none
tlscommon.ServerConfig: ca_sha256=[]
tlscommon.ServerConfig: CertificateSettings: checking certificate keypair
tlscommon.ServerConfig: CertificateSettings: certificate keypair OK.
tlscommon.ServerConfig: CertificateSettings: cert 0 Subject=CN=fleet-server-dev,O=elastic-fleet
	Issuer=CN=localhost,O=elastic-fleet
	IsCA=false
	BasicConstrintsValid=false
	NotBefore=2024-05-29 22:08:32 +0000 UTC
	NotAfter=2034-05-29 22:08:32 +0000 UTC
	Fingerprint=w04CVUAd7Tfnd/AZ/A8YB9bCUAImGeGXBAUPnhEvGgM=
	SAN IP=[]
	SAN DNS=[fleet-server-dev]
	SAN URI=[]
tlscommon.ServerConfig: CertificateAuthorities: certificate_authorities provided.
tlscommon.ServerConfig: CertificateAuthorities: - cert 0 Subject=CN=localhost,O=elastic-fleet
	Issuer=CN=localhost,O=elastic-fleet
	IsCA=true
	BasicConstrintsValid=true
	NotBefore=2024-05-29 22:08:23 +0000 UTC
	NotAfter=2034-05-29 22:08:23 +0000 UTC
	Fingerprint=QHVJrVpzEHBJTTr6F9RW6TnYbj9ckPD2BQ5JphTzh7M=
	SAN IP=[]
	SAN DNS=[localhost]
	SAN URI=[]
```